File: src/packages/sdk/src/cli/render/sort.ts
/**
 * Deterministic ordering for failed cases.
 * Sort by status severity (failed > error > skipped > passed), then by testCaseId asc.
 */

const STATUS_SEVERITY: Record<string, number> = {
	failed: 0,
	error: 1,
	skipped: 2,
	passed: 3,
};

export interface SortableCase {
	status?: string;
	testCaseId?: number;
	[key: string]: unknown;
}

export function sortFailedCases<T extends SortableCase>(cases: T[]): T[] {
	return [...cases].sort((a, b) => {
		const sevA = STATUS_SEVERITY[a.status?.toLowerCase() ?? ""] ?? 4;
		const sevB = STATUS_SEVERITY[b.status?.toLowerCase() ?? ""] ?? 4;
		if (sevA !== sevB) return sevA - sevB;
		const idA = a.testCaseId ?? 0;
		const idB = b.testCaseId ?? 0;
		return idA - idB;
	});
}

File: src/packages/sdk-python/src/evalgate_sdk/cli/render/sort.py
"""Deterministic ordering for failed cases.

Sort by status severity (failed > error > skipped > passed), then by test_case_id asc.

Port of ``cli/render/sort.ts``.
"""

from __future__ import annotations

from typing import Any

STATUS_SEVERITY: dict[str, int] = {
    "failed": 0,
    "error": 1,
    "skipped": 2,
    "passed": 3,
}


def sort_failed_cases(cases: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Sort cases by status severity then test_case_id."""

    def sort_key(c: dict[str, Any]) -> tuple[int, int]:
        status = (c.get("status") or "").lower()
        sev = STATUS_SEVERITY.get(status, 4)
        tid = c.get("test_case_id") or c.get("testCaseId") or 0
        return (sev, tid)

    return sorted(cases, key=sort_key)
